pkg/nvme/initiator.go

```
@@ -23,103 +23,242 @@ const (
-			logrus.WithError(err).Warnf("Failed to connect target, ")
+
+		if i.ControllerName, err = ConnectTarget(transportAddress, transportServiceID, i.SubsystemNQN, i.executor); err != nil {
+			i.logger.WithError(err).Warnf("Failed to connect target")
 			time.Sleep(RetryInterval)
 			continue
 		}
-		i.logger.WithField("controllerName", i.ControllerName)
 		break
 	}
 
 	if i.ControllerName == "" {
-		return fmt.Errorf("failed to start initiator within %d * %vsec retrys", RetryCounts, RetryInterval.Seconds())
+		return fmt.Errorf("failed to start NVMe initiator %s within %d * %vsec retrys", i.Name, RetryCounts, RetryInterval.Seconds())
+	}
+
+	if err := i.loadStartedDeviceInfoWithoutLock(); err != nil {
+		return errors.Wrapf(err, "failed to load device info after starting NVMe initiator %s", i.Name)
+	}
+
+	if err := i.makeEndpoint(); err != nil {
+		return err
 	}
 
+	i.isUp = true
+	i.logger.Infof("Launched NVMe initiator")
+
 	return nil
 }
 
-func (i *Initiator) StopInitiator() error {
-	lock := nsfilelock.NewLockWithTimeout(util.GetHostNamespacePath(HostProc), LockFile, LockTimeout)
-	if err := lock.Lock(); err != nil {
-		return errors.Wrap(err, "failed to lock")
+func (i *Initiator) Stop() error {
+	if i.hostProc != "" {
+		lock := nsfilelock.NewLockWithTimeout(util.GetHostNamespacePath(i.hostProc), LockFile, LockTimeout)
+		if err := lock.Lock(); err != nil {
+			return errors.Wrapf(err, "failed to get file lock for NVMe initiator %s", i.Name)
+		}
+		defer lock.Unlock()
 	}
-	defer lock.Unlock()
 
-	ne, err := util.NewNamespaceExecutor(util.GetHostNamespacePath(HostProc))
-	if err != nil {
+	return i.stopWithoutLock()
+}
+
+func (i *Initiator) stopWithoutLock() error {
+	if err := i.removeEndpoint(); err != nil {
 		return err
 	}
 
-	if err := DisconnectTarget(i.SubsystemNQN, ne.Execute); err != nil {
+	if err := DisconnectTarget(i.SubsystemNQN, i.executor); err != nil {
```

Should we ignore the error if the target does not exist?

---

There is no error when the target does not exists.